vpr/src/backend/physical_netlist_writer.cpp

```diff
@@ -0,0 +1,732 @@
+    VTR_ASSERT(gzclose(file) == Z_OK);
+}
+
+void write_physical_netlist(t_vpr_setup& vpr_setup) {
+    auto& device_ctx = g_vpr_ctx.device();
+
+    // Opening device file
+    gzFile file = gzopen(vpr_setup.FileNameOpts.ArchFile.c_str(), "r");
+    VTR_ASSERT(file != Z_NULL);
+
+    std::vector<uint8_t> output_data;
+    output_data.resize(4096);
+    std::stringstream sstream(std::ios_base::in | std::ios_base::out | std::ios_base::binary);
+    while (true) {
+        int ret = gzread(file, output_data.data(), output_data.size());
+        VTR_ASSERT(ret >= 0);
+        if (ret > 0) {
+            sstream.write((const char*)output_data.data(), ret);
+            VTR_ASSERT(sstream);
+        } else {
+            VTR_ASSERT(ret == 0);
+            int error;
+            gzerror(file, &error);
+            VTR_ASSERT(error == Z_OK);
+            break;
+        }
+    }
+
+    VTR_ASSERT(gzclose(file) == Z_OK);
+
+    sstream.seekg(0);
+    kj::std::StdInputStream istream(sstream);
+
+    // Reader options
+    capnp::ReaderOptions reader_options;
+    reader_options.nestingLimit = std::numeric_limits<int>::max();
+    reader_options.traversalLimitInWords = std::numeric_limits<uint64_t>::max();
+
+    capnp::InputStreamMessageReader message_reader(istream, reader_options);
+
+    auto device_reader = message_reader.getRoot<DeviceResources::Device>();
+
+    ::capnp::MallocMessageBuilder message;
+    PhysicalNetlist::PhysNetlist::Builder physical_netlist = message.initRoot<PhysicalNetlist::PhysNetlist>();
+    physical_netlist.setPart(vpr_setup.device_layout);
+
+    ic::InterchangeWriterVisitor visitor(&device_ctx.arch->strings, device_reader, physical_netlist);
+    NetlistWalker nl_walker(visitor);
+    nl_walker.walk();
+
+    std::string circuit_name = vtr::split(vpr_setup.FileNameOpts.CircuitFile, ".")[0];
+    std::string phys_netlist_filename = circuit_name + ".phys";
+    write_message(message, phys_netlist_filename);
+}
+
+namespace ic {
+
+static t_route_segment get_bel_pin_segment(e_route_type type, size_t site, size_t bel, size_t pin) {
+    t_route_segment bel_pin;
+    bel_pin.type = type;
+    bel_pin.site = site;
+    bel_pin.bel = bel;
+    bel_pin.pin = pin;
+
+    return bel_pin;
+}
+
+static t_route_segment get_site_pin_segment(e_route_type type, size_t site, size_t pin) {
+    t_route_segment site_pin;
+    site_pin.type = type;
+    site_pin.site = site;
+    site_pin.pin = pin;
+
+    return site_pin;
+}
+
+static t_route_segment get_site_pip_segment(e_route_type type, size_t site, size_t bel, size_t pin) {
+    t_route_segment site_pip;
+    site_pip.type = type;
+    site_pip.site = site;
+    site_pip.bel = bel;
+    site_pip.pin = pin;
+
+    return site_pip;
+}
+
+static t_route_segment get_pip_segment(e_route_type type, size_t tile, size_t wire0, size_t wire1) {
+    t_route_segment pip;
+    pip.type = type;
+    pip.tile = tile;
+    pip.wire0 = wire0;
+    pip.wire1 = wire1;
+
+    return pip;
+}
```

Comment:
This is unnecessary. A single constructor with default arguments would be more than enough.

Reply:
Given that the route branches can have different content, it is not trivial to generalize it. In the sense that, given that the type of the input is always `size_t` we cannot differentiate what case is what and what we are dealing with. I think it is better to leave these explicit